File: src/free/java/com/juick/android/NotificationManager.java
/*
 * Copyright (C) 2008-2020, Juick
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.juick.android;

import android.content.Context;
import android.util.Log;

import com.here.oksse.OkSse;
import com.here.oksse.ServerSentEvent;
import com.juick.App;
import com.juick.BuildConfig;
import com.juick.api.model.Post;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;

public class NotificationManager {

    private static final String TAG = NotificationManager.class.getSimpleName();

    private Context context;

    private OkHttpClient es;

    public NotificationManager(Context context) {
        this.context = context;
    }
    public void onResume() {
        if (es != null) return;
        es = Utils.getSSEFactory()
                .readTimeout(0, TimeUnit.SECONDS).build();
        Request request = new Request.Builder()
                .url(BuildConfig.EVENTS_ENDPOINT)
                .build();
        OkSse sse = new OkSse(es);
        sse.newServerSentEvent(request, new ServerSentEvent.Listener() {
            @Override
            public void onOpen(ServerSentEvent sse, okhttp3.Response response) {
                Log.d(TAG, "Event listener opened");
            }

            @Override
            public void onMessage(ServerSentEvent sse, String id, String event, String message) {
                Log.d(TAG, "event received: " + event);
                if (event.equals("msg")) {
                    try {
                        Post reply = App.getInstance().getJsonMapper().readValue(message, Post.class);
                        App.getInstance().getNewMessage().postValue(reply);
                    } catch (IOException e) {
                        Log.d(TAG, "JSON exception: " + e.getMessage());
                    }
                }
            }

            @Override
            public void onComment(ServerSentEvent sse, String comment) {

            }

            @Override
            public boolean onRetryTime(ServerSentEvent sse, long milliseconds) {
                return true;
            }

            @Override
            public boolean onRetryError(ServerSentEvent sse, Throwable throwable, okhttp3.Response response) {
                return true;
            }

            @Override
            public void onClosed(ServerSentEvent sse) {
            }

            @Override
            public Request onPreRetry(ServerSentEvent sse, Request originalRequest) {
                return originalRequest;
            }
        });
    }
}

File: src/free/java/com/juick/android/NotificationManager.kt
/*
 * Copyright (C) 2008-2024, Juick
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.juick.android

import accountData
import android.util.Log
import com.here.oksse.OkSse
import com.here.oksse.ServerSentEvent
import com.juick.App
import com.juick.BuildConfig
import com.juick.android.Utils.eventsFactory
import com.juick.api.model.Post
import kotlinx.coroutines.flow.update
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit

class NotificationManager {
    private lateinit var events: ServerSentEvent
    fun onResume() {
        val es = eventsFactory!!
            .readTimeout(0, TimeUnit.SECONDS).build()
        val uri = Utils.buildUrl(BuildConfig.EVENTS_ENDPOINT)
        val hash = App.instance.accountData
        if (hash.isNotEmpty()) {
            uri.appendQueryParameter("hash", hash)
        }
        val request = Request.Builder()
            .url(uri.build().toString())
            .build()
        val sse = OkSse(es)
        events = sse.newServerSentEvent(request, object : ServerSentEvent.Listener {
            override fun onOpen(sse: ServerSentEvent, response: Response) {
                Log.d(TAG, "Event listener opened")
            }

            override fun onMessage(
                sse: ServerSentEvent,
                id: String?,
                event: String,
                message: String
            ) {
                Log.d(TAG, "event received: $event")
                if (event == "msg") {
                    try {
                        val reply: Post =
                            App.instance.jsonMapper.decodeFromString<Post>(message)
                        App.instance.messages.update { listOf(reply) }
                    } catch (e: IOException) {
                        Log.d(TAG, "JSON exception: " + e.message)
                    }
                }
            }

            override fun onComment(sse: ServerSentEvent, comment: String) {}
            override fun onRetryTime(sse: ServerSentEvent, milliseconds: Long): Boolean {
                return true
            }

            override fun onRetryError(
                sse: ServerSentEvent,
                throwable: Throwable,
                response: Response?
            ): Boolean {
                return true
            }

            override fun onClosed(sse: ServerSentEvent) {
                Log.d(TAG, "Event listener closed")
            }
            override fun onPreRetry(sse: ServerSentEvent, originalRequest: Request): Request {
                return originalRequest
            }
        })
    }

    fun onPause() {
        events.close()
    }

    companion object {
        private val TAG = NotificationManager::class.java.simpleName
    }
}